File: my-folder/0216-combination-sum-iii/solution.java
class Solution {
    public List<List<Integer>> combinationSum3(int k, int n) {
     
        
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> tmp = new ArrayList<>();
        
        backtrack(res, tmp, 1, k, n);
        
        return res;
    }
    
    private void backtrack(List<List<Integer>> res, List<Integer> tmp, int start, int k, int n) {
        if (n < 0) return;
        // base case
        if (tmp.size() == k && n == 0) {
            res.add(new ArrayList<Integer>(tmp));
            return;
        }
        
        for (int i = start; i <= 9; i++) {
            tmp.add(i);
            
            backtrack(res, tmp, i+1, k, n - i);
            
            tmp.remove(tmp.size() - 1);
        }
    }
    
}

File: my-folder/0216-combination-sum-iii/solution.py
class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        res = []
        def traverse(current_res, s, i):
            if len(current_res)==k and s == n:
                res.append(current_res)
            elif i<=9 and len(current_res)<k and i<=n-s:
                traverse(current_res+[i], s+i, i+1)
                traverse(current_res, s, i+1)
            # print(current_res)
        traverse([], 0, 1)
        return res
            
